File: LongChallenge_2019_04_MAXREM.hs
-- code_report Solution
-- Video Link: https://youtu.be/JT3AYsVtYKM
-- Problem Link: https://www.codechef.com/APRIL19B/problems/MAXREM/

import Control.Monad
import Data.List
import Data.Maybe
import Data.Char 

import qualified Data.ByteString.Char8 as B
readInt = readLn :: IO Int
readInts = map ( fst . fromJust . B.readInt ) . B.words <$> B.getLine

unique = map head . group . sort

solve :: [Int] -> Int
solve xs = if length ys == 1 then 0 else last $ init ys
    where ys = unique xs

main :: IO ()
main = do
    _ <- readInt
    v <- readInts
    print $ solve v 

File: LongChallenge_2019_04_MAXREM.cpp
// code_report Solution
// Video Link: https://youtu.be/JT3AYsVtYKM
// Problem Link: https://www.codechef.com/APRIL19B/problems/MAXREM/

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

auto solve(vector<int> v) -> int {
    sort(v.begin(), v.end());
    v.erase(unique(v.begin(), v.end()), v.end());
    return v.size() == 1 ? 0 : v[v.size() - 2];
}

int main() {
    int n; cin >> n;
    vector<int> v(n);
    for (int i = 0; i < n; ++i) cin >> v[i];
    cout << solve(v) << endl;
    return 0;
}
